core/src/banking_stage.rs

```diff
@@ -1853,57 +1866,100 @@ impl BankingStage {
         my_pubkey: &Pubkey,
         next_leader: Option<Pubkey>,
         banking_stage_stats: &BankingStageStats,
-    ) -> Vec<usize> {
+    ) -> usize {
+        let (bigger_queue, smaller_queue, is_unprocessed_packets_bigger) =
+            if retryable_packets.len() > unprocessed_packets.len() {
+                (retryable_packets, unprocessed_packets, false)
+            } else {
+                (unprocessed_packets, retryable_packets, true)
+            };
+        let original_unprocessed_packets_len = bigger_queue.len() + smaller_queue.len();
+
         // Check if we are the next leader. If so, let's not filter the packets
         // as we'll filter it again while processing the packets.
         // Filtering helps if we were going to forward the packets to some other node
-        if let Some(leader) = next_leader {
-            if leader == *my_pubkey {
-                return deserialized_packet_batch
-                    .keys()
-                    .cloned()
-                    .collect::<Vec<usize>>();
+        let will_still_be_leader = next_leader
+            .map(|next_leader| next_leader == *my_pubkey)
+            .unwrap_or(false);
+        let should_filter_unprocessed_packets = !will_still_be_leader && bank.is_some();
+
+        if !should_filter_unprocessed_packets {
+            // If we shouldn't filter the packets, then simply merge the
+            // retryable queue and the unprocessed packets queue so we don't
+            // lose any packets for forwarding or for the next time we're
+            // processing transactions
+            for packet in smaller_queue.iter() {
+                bigger_queue.push(packet.clone());
             }
-        }
+        } else {
+            // If `should_filter_unprocessed_packets` is true, then the bank
+            // must be `Some`
+            let bank = bank.as_ref().unwrap();
+            let mut unprocessed_packet_conversion_time =
+                Measure::start("unprocessed_packet_conversion");
+
+            // Retain elements of larger queue, only perf benefit here would be
+            // if retain was optimized, which currently is is not.
+            bigger_queue.retain(|deserialized_packet| {
+                Self::transaction_from_deserialized_packet(
+                    deserialized_packet,
+                    &bank.feature_set,
+                    bank.vote_only_bank(),
+                    bank.as_ref(),
+                )
+                .is_some()
+            });
 
-        let mut unprocessed_packet_conversion_time =
-            Measure::start("unprocessed_packet_conversion");
-        let (transactions, transaction_to_packet_indexes) = Self::transactions_from_packets(
-            deserialized_packet_batch,
-            &bank.feature_set,
-            bank.vote_only_bank(),
-            bank.as_ref(),
-        );
-        unprocessed_packet_conversion_time.stop();
+            for deserialized_packet in smaller_queue.iter() {
```

Comment:
we can share more code if we `retain` both queues in this case, then just `extend` `bigger` with `smaller` unconditionally afterwards. avoid the `swap` and the `clear` as a bonus

Reply:
Hmm right now `retain` is arguably more work because it requires draining and recreating the MinMaxHeap. It seems unnecessary to `retain` both and incur the clone on both heaps.

I'm wondering if we should instead be calling `retain` on the smaller heap and extend it with the bigger heap, thus avoiding the `drain` on the bigger heap

Reply:
I'll take your word for it.  Thought we were retaining both anyway, but I can't find the commit I originally commented on and the current version doesn't do that. So :+1: